sticker-utils/src/subcommands/finetune.rs

```diff
@@ -1,20 +1,239 @@
+use std::collections::BTreeMap;
 use std::fs::File;
 
 use clap::{App, Arg, ArgMatches};
-use conllx::io::{ReadSentence, Reader};
+use failure::Fallible;
+use indicatif::ProgressStyle;
+use ordered_float::NotNan;
 use serde_yaml;
-use stdinout::{Input, OrExit};
+use stdinout::OrExit;
 use sticker::config::{Config, TomlRead};
+use sticker::dataset::{ConllxDataSet, DataSet};
 use sticker::encoders::Encoders;
+use sticker::input::vectorizer::WordPieceVectorizer;
+use sticker::input::WordPieceTokenizer;
+use sticker::lr::{ExponentialDecay, LearningRateSchedule, PlateauLearningRate};
+use sticker::model::BertModel;
+use sticker::optimizers::{AdamW, AdamWConfig};
+use sticker::util::seq_len_to_mask;
+use tch::nn::VarStore;
+use tch::{self, Device, Kind};
 
+use crate::progress::ReadProgress;
+use crate::save::{BestEpochSaver, CompletedUnit, Save};
 use crate::traits::{StickerApp, DEFAULT_CLAP_SETTINGS};
 
+const BATCH_SIZE: &str = "BATCH_SIZE";
 const CONFIG: &str = "CONFIG";
-static TRAIN_DATA: &str = "TRAIN_DATA";
+const GPU: &str = "GPU";
+const FINETUNE_EMBEDS: &str = "FINETUNE_EMBEDS";
+const INITIAL_LR_CLASSIFIER: &str = "INITIAL_LR_CLASSIFIER";
+const INITIAL_LR_ENCODER: &str = "INITIAL_LR_ENCODER";
+const LABEL_SMOOTHING: &str = "LABEL_SMOOTHING";
+const LR_DECAY_RATE: &str = "LR_DECAY_RATE";
+const LR_PATIENCE: &str = "LR_PATIENCE";
+const LR_SCALE: &str = "LR_SCALE";
+const MAX_LEN: &str = "MAX_LEN";
+const PATIENCE: &str = "PATIENCE";
+const PRETRAINED_MODEL: &str = "PRETRAINED_MODEL";
+const TRAIN_DATA: &str = "TRAIN_DATA";
+const VALIDATION_DATA: &str = "VALIDATION_DATA";
+const WARMUP: &str = "WARMUP";
+const WEIGHT_DECAY: &str = "WEIGHT_DECAY";
+
+pub struct LrSchedule {
+    pub initial_lr_encoder: NotNan<f32>,
+    pub initial_lr_classifier: NotNan<f32>,
+    pub lr_decay_rate: NotNan<f32>,
+    pub lr_scale: NotNan<f32>,
+    pub lr_patience: usize,
+    pub warmup_steps: usize,
+}
 
 pub struct FinetuneApp {
+    batch_size: usize,
     config: String,
-    train_data: Option<String>,
+    device: Device,
+    finetune_embeds: bool,
+    max_len: Option<usize>,
+    label_smoothing: Option<f64>,
+    lr_schedule: LrSchedule,
+    patience: usize,
+    pretrained_model: String,
+    saver: BestEpochSaver<f32>,
+    train_data: String,
+    validation_data: String,
+    weight_decay: f64,
+}
+
+pub struct LearningRateSchedules {
+    pub classifier: PlateauLearningRate<ExponentialDecay>,
+    pub encoder: PlateauLearningRate<ExponentialDecay>,
+}
```

Comment:
Why is it not possible to use any other schedules?

Reply:
Good question! We had the same issue in sticker1. The problem is that we have not found a clean way yet with `clap` to have/show options dependent on some other option. All these LR schedules have different hyperparameters and we'd want to be able to say something like `--lr-schedule expdecay` and only have the options relevant to exponential decay available. You can hack your way around it by not permitting other options, but the UI gets pretty ugly, because you cannot show options dependent on another option in groups.

Reply:
💡 Thanks for clarifying!

Reply:
Maybe it could move to a config file at some point.

Reply:
Indeed, it seems like a reasonable default and you can even disable one of the two by setting the right hyper parameters ;).